eldritch/events.py

```diff
@@ -593,24 +575,172 @@ def resolve(self, state):
   def is_resolved(self):
     return self.kept is not None
 
   def start_str(self):
-    return f"{self.character.name} draws {self.draw_count} cards from the {self.deck} deck"
+    return ""
 
   def finish_str(self):
-    return f"{self.character.name} keeps " + ", ".join(self.kept)
+    return f"{self.character} kept " + ", ".join(self.kept)
+
+def Draw(character, deck, draw_count, prompt="Choose a card", target_type=None):
+  cards = DrawItems(character, deck, draw_count, target_type=target_type)
+  keep = KeepDrawn(character, cards, prompt)
+  return Sequence([cards, keep])
 
+class PurchaseDrawn(Event):
+
+  def __init__(self, character, draw: DrawItems,
+               discount_type="fixed", discount=0, keep_count=1, target_type=None, prompt="Buy items?"):
+    assert discount_type in {"fixed", "rate"}
+    self.character = character
+    self.prompt = prompt
+    self.keep_count = keep_count
+    self.discount_type = discount_type
+    self.discount = discount
+    self.draw = draw
+    self.drawn = None
+    self.choice = None
+    self.kept = []
+    self.prices = None
+    self.target_type = target_type
+    self.resolved = False
+
+  def resolve(self, state):
+    if self.resolved:
+      # This should never happen??
+      return True
+    if self.drawn is None:
+      assert self.draw.is_resolved()
+      self.drawn = self.draw.drawn
+
+    if self.choice is not None:
+      if not self.choice.is_resolved():  # This should never happen
+        return False
+      if self.choice.choices[self.choice.choice_index] == "Nothing":
+        self.resolved = True
+        for card in self.drawn:
+          getattr(state, self.draw.deck).append(card)
+        return True
+      kept_card = self.drawn.pop(self.choice.choice_index)
+      cost = self.prices.pop(self.choice.choice_index)
+      self.kept.append(self.choice.choices[self.choice.choice_index])
+      assert cost <= self.character.dollars
+      self.character.dollars -= cost
+      self.character.possessions.append(kept_card)
+      self.keep_count -= 1
+
+    if self.keep_count == 0:
+      for card in self.drawn:
+        getattr(state, self.draw.deck).append(card)
+      self.resolved = True
+      return True
+
+    # self.keep_count > 0
+    choices = []
+    available = []
+    unavailable = []
+    self.prices = []
+    for card in self.drawn:
+      price = self.discounted_price(card)
+      if price <= self.character.dollars:
+        available.append(card)
+        self.prices.append(price)
+        choices.append(f"{card.name} for ${price}")
+      else:
+        getattr(state, self.draw.deck).append(card)
+        unavailable.append(f"{card.name} for ${price}")
+    self.drawn = available
+    choices.append("Nothing")
+
+    if unavailable:
+      could_not_afford = " (Could not afford {})".format(",".join(unavailable))
+    else:
+      could_not_afford = ""
+    if available:
+      self.choice = CardChoice(self.character, self.prompt + could_not_afford, choices)
```

Comment:
The choices for CardChoice have to be the card names (this is so the UI knows which cards to show the user). We probably need to add annotations to CardChoice so that they can see the final price.

Reply:
Reverted, and created an issue so we remember that we need to do it.